overlord/snapstate/handlers.go

```diff
@@ -1384,21 +1471,49 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	snapst.Classic = oldClassic
 	snapst.RefreshInhibitedTime = oldRefreshInhibitedTime
 	snapst.CohortKey = oldCohortKey
+	snapst.LastActiveDisabledServices = oldLastActiveDisabledServices
 
 	newInfo, err := readInfo(snapsup.InstanceName(), snapsup.SideInfo, 0)
 	if err != nil {
 		return err
 	}
 
-	// we need to undo potential changes to current snap configuration (e.g. if modified by post-refresh/install/configure hooks
-	// as part of failed refresh/install) by restoring the configuration of "old current".
+	// we need to undo potential changes to current snap configuration (e.g. if
+	// modified by post-refresh/install/configure hooks as part of failed
+	// refresh/install) by restoring the configuration of "old current".
+	// similarly, we need to re-save the disabled services if there is a
+	// revision for us to go back to, see comment below for full explanation
 	if len(snapst.Sequence) > 0 {
 		if err = config.RestoreRevisionConfig(st, snapsup.InstanceName(), oldCurrent); err != nil {
 			return err
 		}
+
+		// unlock state while we talk to systemd
+		st.Unlock()
+		defer st.Lock()
+
+		// get the currently disabled services and add them to
+		// snapst.LastActiveDisabledServices because if we completed a successful
+		// doLinkSnap (hence we are in the undo handler), then we already disabled
+		// the services and deleted currently existing services from the state
+		// during doLinkSnap, but now we will need that information again when we go
+		// to link the old version to prevent accidental enabling of disabled
+		// services on a failed revert/refresh
+		disabledServices, err := m.queryDisabledServices(newInfo, NewTaskProgressAdapterUnlocked(t))
```

Comment:
This should probably be refactored a bit to make the tracking of st.Unlock()/st.Lock() a bit easier, eg. like so:
```
st.Unlock()
disabledServices, err := m.queryDisabledServices(newInfo, NewTaskProgressAdapterUnlocked(t))
st.Lock()

Reply:
As we discussed on IRC, what I have is a bit more readable and safer against future refactors, so I will leave this as is for now.